cpp/src/c_api/graph_sg.cpp

```diff
@@ -15,21 +15,233 @@
  */
 
 #include <cugraph_c/graph.h>
+#include <c_api/abstract_functor.hpp>
+#include <c_api/array.hpp>
+#include <c_api/error.hpp>
+#include <c_api/graph.hpp>
+
+#include <cugraph/detail/utility_wrappers.hpp>
+#include <cugraph/graph_functions.hpp>
+#include <cugraph/visitors/generic_cascaded_dispatch.hpp>
+
+#include <raft/handle.hpp>
+
+namespace cugraph {
+namespace c_api {
+
+struct create_graph_functor : public abstract_functor {
+  raft::handle_t const& handle_;
+  cugraph_graph_properties_t const* properties_;
+  c_api::cugraph_type_erased_device_array_t* src_;
+  c_api::cugraph_type_erased_device_array_t* dst_;
+  c_api::cugraph_type_erased_device_array_t* weights_;
+  bool_t renumber_;
+  bool_t check_;
+  data_type_id_t edge_type_;
+  c_api::cugraph_graph_t* result_{};
+
+  create_graph_functor(raft::handle_t const& handle,
+                       cugraph_graph_properties_t const* properties,
+                       c_api::cugraph_type_erased_device_array_t* src,
+                       c_api::cugraph_type_erased_device_array_t* dst,
+                       c_api::cugraph_type_erased_device_array_t* weights,
+                       bool_t renumber,
+                       bool_t check,
+                       data_type_id_t edge_type)
+    : abstract_functor(),
+      properties_(properties),
+      handle_(handle),
+      src_(src),
+      dst_(dst),
+      weights_(weights),
+      renumber_(renumber),
+      check_(check),
+      edge_type_(edge_type)
+  {
+  }
+
+  template <typename vertex_t,
+            typename edge_t,
+            typename weight_t,
+            bool store_transposed,
+            bool multi_gpu>
+  void operator()()
+  {
+    if constexpr (multi_gpu || !cugraph::is_candidate<vertex_t, edge_t, weight_t>::value) {
+      unsupported();
+    } else {
+      if (check_) {
+        // FIXME:  Need an implementation here.
+      }
+
+      vertex_t num_vertices{};
+      renumber_meta_t<vertex_t, edge_t, multi_gpu> meta;
+
+      rmm::device_uvector<vertex_t>* number_map =
+        new rmm::device_uvector<vertex_t>(0, handle_.get_stream());
+
+      if (renumber_) {
```

Comment:
I think we'd better just call create_graph_from_edgelist instead of separately calling renumber_edgelist and graph constructor.

Reply:
I don't have a `device_uvector` as input.  I was trying to avoid copying the data... although I think it's probably better to do the copy.  I'll change this.

Reply:
Yeah... maybe copy at this moment (add a FIXME statement), and later, we may create another version of ```create_graph_from_edgelist``` taking pointers if turns out to be necessary.

Copying can have two issues: 1) memory footprint & 2) cost of copying.

1) & 2) may not be a huge issue if the data is coming from end-users and we are not allowed to destroy the data (as we need to sort/shuffle the edge list inside create_graph_from_edgelist and if we are not allowed to modify, then we should create a copy anyways).

Copy can be wasteful in both memory & computing if we are allowed to modify/destroy the input data, but AFAIK, this is not a case if we are dealing with the data coming from cugraph python users (if users are just passing file names... then, yes, we can modify the edge list). If we encounter this use case, then, I think we'd better create a second version of create_graph_from_edgelist.

Reply:
Created #1947

Reply:
Currently we can't destroy the data, so copying is probably reasonable.  I modified the code to do the copy.